format/spec.md

```diff
@@ -241,7 +245,14 @@ Struct evolution requires the following rules for default values:
 
 #### Column Projection
 
-Columns in Iceberg data files are selected by field id. The table schema's column names and order may change after a data file is written, and projection must be done using field ids. If a field id is missing from a data file, its value for each row should be `null`.
+Columns in Iceberg data files are selected by field id. The table schema's column names and order may change after a data file is written, and projection must be done using field ids.
+
+Values for field ids which are not present in a data file must be resolved according the following rules:
+
+* Return the value from partition metadata if an [Identity Transform](#partition-transforms) exists for the field and the partition value is present in the `partition` struct on `data_file` object in the manifest. 
```

Comment:
if a new column projection is defined (eg `day(order_time)`), can the query engine derive value for that field from the `order_time` column?

Reply:
Transforms are one way, with the exception of `identity` that does not modify the value. No other partition transforms support recovering the original value.

If you have a case where the engine is deriving a value that is stored in metadata (which I _think_ is what you're asking) then you can use the metadata value as long as you know that the transform the engine is projecting exactly matches the Iceberg transform.

Reply:
I'm not sure I fully understand the question but if order_time is of type date, then in this special case the transform is effectively an identity transform.

Reply:
I was thinking about situation like this (using Trino syntax)

```
-- create a table with some data
CREATE TABLE t AS SELECT 123 AS a;

-- add new partitioning column
ALTER TABLE t SET PROPERTIES partitioning = ARRAY['truncate(a, 10)'];
```

Now we have two fields: the `a` data column and `a_trunc` projected column.
Trino doesn't provide a way to query for `a_trunc` column directly.
However, if it did, the value for `a_trunc` could be derived from the data.

Reply:
It seems for this case we probably want a subsection of the specification or an implementation note for projecting partition values?  Maybe that would be a good follow-up (I'm not sure which engines actually allow projecting partition values since they aren't technically part of the schema.)

Reply:
Yes, agreed that partition columns are not part of the schema, so don't show up in table's relational model.
However, the document here talks about reading "field ids" and partition columns are also fields with ids. Or is this section supposed to be only about fields that are plain data fields?

Reply:
I read it as at least initially only about plain data, since for transform partition values these would never actually be in data file, and that is what this topic currently discusses.  I think adding a subsection about projecting partition values if necessary to this section as a follow-up could make sense.

Reply:
This is allowed, particularly for filters where you can use Iceberg's expression library to filter by partition values directly. You can also use partition values in aggregations (which is under development).

I don't think that changes what is said here. In this case, the value from metadata should be used if it is present.

Reply:
> the value from metadata should be used if it is present.

makes sense, agreed

but if it's not present -- why require to return null? we could require the query engine to compute the value based on the source columns present in the data file

Reply:
@findepi, the spec originally stated that all columns that were not present in a file must be interpreted as `null` because the file is the source of truth for row data -- if a column was not present then it must not have existed at the time the file was written and must default to `null` because that was the only possible default for new columns.

This is how identity partition columns were always required to be written because omitting them would create a conflict between the partition value in metadata and the source of truth in the file. Now, the requirement to write all columns is more explicitly stated above.

Over time, we added support for reading Hive files that were missing the identity partitioned columns. This now documents how to do that in the read path but we didn't drop the requirement to write all columns in the write path.

In addition, we've now added column defaults for v3 and didn't realize that this section conflicted with the `initial-default` behavior.

This is just stating more clearly what the expected behavior is and I think it's a good change because it is clear.